builtin/logical/totp/path_keys.go

```diff
@@ -0,0 +1,400 @@
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+func (b *backend) pathKeyDelete(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	err := req.Storage.Delete("key/" + data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (b *backend) pathKeyRead(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	key, err := b.Key(req.Storage, data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+	if key == nil {
+		return nil, nil
+	}
+
+	// Translate algorithm back to string
+	algorithm := key.Algorithm.String()
+
+	// Return values of key
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"issuer":       key.Issuer,
+			"account_name": key.AccountName,
+			"period":       key.Period,
+			"algorithm":    algorithm,
+			"digits":       key.Digits,
+		},
+	}, nil
+}
+
+func (b *backend) pathKeyList(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	entries, err := req.Storage.List("key/")
+	if err != nil {
+		return nil, err
+	}
+
+	return logical.ListResponse(entries), nil
+}
+
+func (b *backend) pathKeyCreate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
+	generate := data.Get("generate").(bool)
+	keyString := data.Get("key").(string)
+	issuer := data.Get("issuer").(string)
+	accountName := data.Get("account_name").(string)
+	period := data.Get("period").(int)
+	algorithm := data.Get("algorithm").(string)
+	digits := data.Get("digits").(int)
+	skew := data.Get("skew").(int)
+	qrSize := data.Get("qr_size").(int)
+	keySize := data.Get("key_size").(int)
+	inputURL := data.Get("url").(string)
+
+	if generate == true {
+		if keyString != "" {
+			return logical.ErrorResponse("a key should not be passed if generate is true"), nil
+		}
+		if inputURL != "" {
+			return logical.ErrorResponse("a url should not be passed if generate is true"), nil
+		}
+	}
+
+	// Read parameters from url if given
+	if inputURL != "" {
+		//Parse url
+		urlObject, err := url.Parse(inputURL)
+		if err != nil {
+			return logical.ErrorResponse("an error occured while parsing url string"), nil
+		}
+
+		//Set up query object
+		urlQuery := urlObject.Query()
+		path := strings.TrimPrefix(urlObject.Path, "/")
+		index := strings.Index(path, ":")
+
+		//Read issuer
+		urlIssuer := urlQuery.Get("issuer")
+		if urlIssuer != "" {
+			issuer = urlIssuer
+		} else {
+			if index != -1 {
+				issuer = path[:index]
+			}
+		}
+
+		//Read account name
+		if index == -1 {
+			accountName = path
+		} else {
+			accountName = path[index+1:]
+		}
+
+		//Read key string
+		keyString = urlQuery.Get("secret")
+
+		//Read period
+		periodQuery, err := strconv.Atoi(urlQuery.Get("period"))
+		if err == nil {
+			period = periodQuery
+		}
+
+		//Read digits
+		digitsQuery, err := strconv.Atoi(urlQuery.Get("digits"))
+		if err == nil {
```

Comment:
This should probably be `if err != nil` and error out if digits can't be parsed. Same with the period above.